nex/crowdsale.py

```diff
@@ -0,0 +1,207 @@
+from boa.interop.Neo.Blockchain import GetHeight
+from boa.interop.Neo.Runtime import CheckWitness
+from boa.interop.Neo.Action import RegisterAction
+from boa.interop.Neo.Storage import Get, Put
+from boa.builtins import concat
+from nex.token import *
+from nex.txio import get_asset_attachments
+
+# OnInvalidKYCAddress = RegisterAction('invalid_registration', 'address')
+OnKYCRegister = RegisterAction('kyc_registration', 'address')
+OnTransfer = RegisterAction('transfer', 'addr_from', 'addr_to', 'amount')
+OnRefund = RegisterAction('refund', 'addr_to', 'amount')
+
+
+def kyc_register(ctx, args):
+    """
+
+    :param args:list a list of addresses to register
+    :param token:Token A token object with your ICO settings
+    :return:
+        int: The number of addresses to register for KYC
+    """
+    ok_count = 0
+
+    if CheckWitness(TOKEN_OWNER):
+
+        for address in args:
+
+            if len(address) == 20:
+
+                kyc_storage_key = concat(KYC_KEY, address)
+                Put(ctx, kyc_storage_key, True)
+
+                OnKYCRegister(address)
+                ok_count += 1
+
+    return ok_count
+
+
+def kyc_status(ctx, args):
+    """
+    Gets the KYC Status of an address
+
+    :param args:list a list of arguments
+    :return:
+        bool: Returns the kyc status of an address
+    """
+
+    if len(args) > 0:
+        addr = args[0]
+
+        kyc_storage_key = concat(KYC_KEY, addr)
+
+        return Get(ctx, kyc_storage_key)
+
+    return False
+
+
+def perform_exchange(ctx):
+    """
+
+     :param token:Token The token object with NEP5/sale settings
+     :return:
+         bool: Whether the exchange was successful
+     """
+
+    attachments = get_asset_attachments()  # [receiver, sender, neo, gas]
+
+    # this looks up whether the exchange can proceed
+    exchange_ok = can_exchange(ctx, attachments, False)
+
+    if not exchange_ok:
+        # This should only happen in the case that there are a lot of TX on the final
+        # block before the total amount is reached.  An amount of TX will get through
+        # the verification phase because the total amount cannot be updated during that phase
+        # because of this, there should be a process in place to manually refund tokens
+        if attachments[2] > 0:
+            OnRefund(attachments[1], attachments[2])
+        # if you want to exchange gas instead of neo, use this
+        # if attachments.gas_attached > 0:
+        #    OnRefund(attachments.sender_addr, attachments.gas_attached)
```

Comment:
Awesome work on this :) I found a potential bug though

Although commented this should be 
if attachments[3] > 0:
           OnRefund(attachments[1], attachments[3])